src/ParallelAlgorithms/Amr/Criteria/Loehner.hpp

```diff
@@ -0,0 +1,269 @@
+ */
+template <size_t Dim, typename TensorTags>
+class Loehner : public Criterion {
+ public:
+  struct VariablesToMonitor {
+    using type = std::vector<std::string>;
+    static constexpr Options::String help = {
+        "The tensors to monitor for h-refinement."};
+    static size_t lower_bound_on_size() { return 1; }
+  };
+  struct RelativeTolerance {
+    using type = double;
+    static constexpr Options::String help = {
+        "If any tensor component has a second derivative magnitude above this "
+        "value times the max of the absolute tensor component over the "
+        "element, the element will be h-refined in that direction. "
+        "Set to 0 to disable."};
+    static double lower_bound() { return 0.; }
+  };
+  struct AbsoluteTolerance {
+    using type = double;
+    static constexpr Options::String help = {
+        "If any tensor component has a second derivative magnitude above this "
+        "value, the element will be h-refined in that direction. "
+        "Set to 0 to disable."};
+    static double lower_bound() { return 0.; }
+  };
+  struct CoarseningFactor {
+    using type = double;
+    static constexpr Options::String help = {
+        "Factor applied to both relative and absolute tolerance to trigger "
+        "h-coarsening. Set to 0 to disable h-coarsening altogether. "
+        "Set closer to 1 to trigger h-coarsening more aggressively. "
+        "Values too close to 1 risk that coarsened elements will immediately "
+        "trigger h-refinement again. A reasonable value is 1/3."};
+    static double lower_bound() { return 0.; }
+    static double upper_bound() { return 1.; }
+  };
+
+  using options = tmpl::list<VariablesToMonitor, RelativeTolerance,
+                             AbsoluteTolerance, CoarseningFactor>;
+
+  static constexpr Options::String help = {
+      "Refine the grid towards the target truncation error"};
+
+  Loehner() = default;
+
+  Loehner(std::vector<std::string> vars_to_monitor, double relative_tolerance,
+          double absolute_tolerance, double coarsening_factor,
+          const Options::Context& context = {});
+
+  /// \cond
+  explicit Loehner(CkMigrateMessage* msg);
+  using PUP::able::register_constructor;
+  WRAPPED_PUPable_decl_template(Loehner);  // NOLINT
+  /// \endcond
+
+  using compute_tags_for_observation_box = tmpl::list<>;
+
+  using argument_tags = tmpl::list<::Tags::DataBox>;
+
+  template <typename DbTagsList, typename Metavariables>
+  std::array<Flag, Dim> operator()(const db::DataBox<DbTagsList>& box,
+                                   Parallel::GlobalCache<Metavariables>& cache,
+                                   const ElementId<Dim>& element_id) const;
+
+  void pup(PUP::er& p) override;
+
+ private:
+  std::vector<std::string> vars_to_monitor_{};
+  double relative_tolerance_{};
+  double absolute_tolerance_{};
+  double coarsening_factor_{};
+};
+
+// Out-of-line definitions
+/// \cond
+
+template <size_t Dim, typename TensorTags>
```

Comment:
not for any of these PRs, but we could conceivably move the function definitions to a `tpp` to reduce compilation of the final executable.

Reply:
That only helps if a separate cpp file includes the `tpp` and instantiates it, not just include the `tpp` in the final exec, right? We could do that for lots of code and perhaps cut down build time by parallelizing the compilation.

Reply:
Yes, that's right. That's what I was thinking, but definitely should first try some of the clang time profiling stuff Larry and Alex Macedo have done.